z3/src/ast.rs

```diff
@@ -381,75 +381,975 @@ impl<'ctx> Ast<'ctx> {
     }
-}
 
-impl<'ctx> PartialEq<Ast<'ctx>> for Ast<'ctx> {
-    fn eq(&self, other: &Ast<'ctx>) -> bool {
-        unsafe { Z3_is_eq_ast(self.ctx.z3_ctx, self.z3_ast, other.z3_ast) }
+    pub fn from_real(ast: &Real<'ctx>) -> Int<'ctx> {
+        Self::new(ast.ctx, unsafe {
+            let guard = Z3_MUTEX.lock().unwrap();
+            Z3_mk_real2int(ast.ctx.z3_ctx, ast.z3_ast)
+        })
+    }
+
+    /// Create a real from an integer.
+    /// This is just a convenience wrapper around
+    /// [`Real::from_int`](struct.Real.html#method.from_int); see notes there
+    pub fn to_real(&self) -> Real<'ctx> {
+        Real::from_int(self)
+    }
+
+    /// Create an integer from a bitvector.
+    ///
+    /// Signed and unsigned version.
+    ///
+    /// # Examples
+    /// ```
+    /// # use z3::{ast, Config, Context, Solver};
+    /// # use z3::ast::Ast;
+    /// # let cfg = Config::new();
+    /// # let ctx = Context::new(&cfg);
+    /// # let solver = Solver::new(&ctx);
+    /// let bv = ctx.named_bitvector_const("x", 32);
+    /// solver.assert(&bv._eq(&ast::BV::from_i64(&ctx, -3, 32)));
+    ///
+    /// let x = ast::Int::from_bv(&bv, true);
+    ///
+    /// assert!(solver.check());
+    /// let model = solver.get_model();
+    ///
+    /// assert_eq!(-3, model.eval(&x).unwrap().as_i64().unwrap());
+    /// ```
+    pub fn from_bv(ast: &BV<'ctx>, signed: bool) -> Int<'ctx> {
+        Self::new(ast.ctx, unsafe {
+            let guard = Z3_MUTEX.lock().unwrap();
+            Z3_mk_bv2int(ast.ctx.z3_ctx, ast.z3_ast, signed)
+        })
+    }
+
+    /// Create a bitvector from an integer.
+    /// This is just a convenience wrapper around
+    /// [`BV::from_int`](struct.BV.html#method.from_int); see notes there
+    pub fn to_ast(&self, sz: u32) -> BV<'ctx> {
+        BV::from_int(self, sz)
     }
+
+    // TODO: this should be on the Ast trait, but I don't know how to return Self<'dest_ctx>.
+    // When I try, it gives the error E0109 "lifetime arguments are not allowed for this type".
+    pub fn translate<'dest_ctx>(&self, dest: &'dest_ctx Context) -> Int<'dest_ctx> {
+        Int::new(dest, unsafe {
+            let guard = Z3_MUTEX.lock().unwrap();
+            Z3_translate(self.ctx.z3_ctx, self.z3_ast, dest.z3_ctx)
+        })
+    }
+
+    varop!(add, Z3_mk_add, Self);
+    varop!(sub, Z3_mk_sub, Self);
+    varop!(mul, Z3_mk_mul, Self);
+    binop!(div, Z3_mk_div, Self);
+    binop!(rem, Z3_mk_rem, Self);
+    binop!(modulo, Z3_mk_mod, Self);
+    binop!(power, Z3_mk_power, Self);
+    unop!(unary_minus, Z3_mk_unary_minus, Self);
+    binop!(lt, Z3_mk_lt, Bool<'ctx>);
+    binop!(le, Z3_mk_le, Bool<'ctx>);
+    binop!(gt, Z3_mk_gt, Bool<'ctx>);
+    binop!(ge, Z3_mk_ge, Bool<'ctx>);
+    // Z3 does support mixing ints and reals in add(), sub(), mul(), div(), and power()
+    //   (but not rem(), modulo(), lt(), le(), gt(), or ge()).
+    // TODO: we could consider expressing this by having a Numeric trait with these methods.
```

Comment:
There's this question about mixing ints and reals in certain methods (which Z3 supports but the current draft of this PR does not)

Reply:
(see the rest of this comment as well, continues to line 527)